### ISSUE TITLE
Creating a template: Present the user with initial block configuration options

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#41060]
When possible we [suggest patterns](https://github.com/WordPress/gutenberg/pull/40034) during the page creation flow. This can help speed up the addition of pages, and provide some creative inspiration. 

It would be good to do the same thing when creating a new template. It's not yet possible to associate a pattern with a template, so we'll need to split this into two parts:

- [x] Offer options to start blank, or with the closest template in the hierarchy (https://github.com/WordPress/gutenberg/pull/37258), for which we can provide a visual preview
- [x] Offer options to start with a contextual pattern, either from the directory or the set bundled with the theme. Blocked by https://github.com/WordPress/gutenberg/pull/45814.

For part one, we'll display something like this after the template is created:

<img width="1196" alt="Screenshot 2022-11-23 at 14 07 51" src="https://user-images.githubusercontent.com/846565/203567338-32a2ee52-f290-459f-9fa1-cf23fda209fe.png">

This will be augmented with additional patterns for part two:

<img width="1197" alt="Screenshot 2022-11-23 at 14 09 35" src="https://user-images.githubusercontent.com/846565/203567557-71d0fee6-4145-4543-8af5-a1897260135f.png">

Somewhat related: https://github.com/WordPress/gutenberg/issues/39308

**Comment by annezazu** [comment]
Might be a duplicate now? https://github.com/WordPress/gutenberg/issues/40238 Would be great to consolidate either way as these feel wildly similar. Happy to close out either.

**Comment by jameskoster** [comment]
I think they're slightly different. #37258 alone would close #40238, but not this one.

**Comment by annezazu** [comment]
Cool I'll close mine out :)

**Comment by carolinan** [comment]
> Start with a contextual pattern from the directory

Please allow this to be from the theme, a pattern from the directory which is not designed for the theme can create a less than ideal experience.

**Comment by jameskoster** [comment]
Shouldn't patterns be agnostic of theme in the long run? Themes including their own patterns feels a bit archaic.

**Comment by mtias** [comment]
Also related #31746.

**Comment by mtias** [comment]
And there's also #31591. @jameskoster can we collect all the related issues and have an easier to follow tracking flow?

**Comment by mtias** [comment]
Also run into #30865. :)

**Comment by carolinan** [comment]
> Shouldn't patterns be agnostic of theme in the long run? Themes including their own patterns feels a bit archaic.

I strongly disagree because I believe there still needs to be room for individuality, uniqueness and creativity.
A theme specifically created for say, a dog rescue, may not want to load patterns built for a pub or dentist.

There is a third aspect that I wish I did not need to bring up, there are many patterns already in the directory that are frankly, too low quality for me to want to feature, and other patterns that could potentially be really good but breaks in the site editor due to broken images or validation issues.

**Comment by jameskoster** [comment]
I think one of the main benefits of patterns is their portability across themes. Bundling them with themes feels a bit like repeating mistakes from the past. 

In any case, if there _are_ patterns bundled with the theme then I think it's fine to display them here. I'll update the OP :)

**Comment by mtias** [comment]
> there are many patterns already in the directory that are frankly, too low quality for me to want to feature, and other patterns that could potentially be really good but breaks in the site editor due to broken images or validation issues

Side note: should these be curated out from the repository? One of the main aspects of patterns is that we don't need to keep them around forever because they have no back-compat issues. While there's obvious room for diversity in taste, if contributors arrive at a consensus that a pattern is low quality, no longer relevant, or any other relevant criteria, it seems fine to tweak or remove them.

**Comment by DaisyOlsen** [comment]
> One of the main aspects of patterns is that we don't need to keep them around forever because they have no back-compat issues.

What would happen if a theme registers a pattern from the directory in theme.json that is later curated out of the directory?

**Comment by jameskoster** [comment]
I quite like the idea of using a modal to contain the options here, mostly because it's portable across the different scenarios in which one might create a template. 

Yesterday I was looking at the template switching flow in the post editor (https://github.com/WordPress/gutenberg/issues/41258#issuecomment-1135977351). In that UI I think we can offer a pathway to create a new template like so:

<details><summary>Old concept</summary>
https://user-images.githubusercontent.com/846565/170250608-72e0b536-475f-4a44-a9cd-82fa9099ff13.mp4
</details>

The patterns in the modal would be pulled in from the theme if it supplies any relevant options, in addition to those found in the pattern directory. 

We can reuse this two-step template creation modal in the site editor too, we'd just need to adjust the wording on the "Duplicate" option.

**Comment by beckej13820** [comment]
Couldn't there be a simple middle ground between what @carolinan is suggesting and using a pattern from the library.

A mechanism where a theme can set the default, but if it doesn't, fallback to a pattern from the directory.

**Comment by jameskoster** [comment]
> A mechanism where a theme can set the default

This already happens to an extent – if you duplicate the current template then most of the time that will mean starting with a design that was originally supplied by the theme. 

This is especially true if you create a new template in the Site Editor. As an example if you create `single-post` and the theme already supplied `single` then you'd have the option to use that as the starting point.

If a user elects to "Start with a pattern" rather than duplicating or starting from scratch, the subsequent modal can prioritise theme-supplied patterns over those pulled in from the directory. I also believe there is a site-level option to not connect to the pattern directory at all, and obviously that would be respected here.

**Comment by jameskoster** [comment]
I think headers and footers will be covered by the work in https://github.com/WordPress/gutenberg/pull/42142, so this issue can concentrate solely on templates.

**Comment by jameskoster** [comment]
Cross-posting a design shared in https://github.com/WordPress/gutenberg/issues/42325#issuecomment-1220515552 because it's very relevant here:

https://user-images.githubusercontent.com/846565/185599910-2f49c633-bd2b-4b8f-be72-bd420bdff2de.mp4

When creating a template the user is presented with a grid of thumbnail previews as options. One of these will be the closest template in the hierarchy, one will be a static "Start blank" link, and the rest will be patterns associated with the chosen template.

**Comment by jameskoster** [comment]
It would be good to try the concept above as both a solution to this issue, and #42325 as well. Template-specific patterns are not on the immediate horizon, but the visual previews and organisation are useful here, as is the option to start blank which is currently missing from the experience.

**Comment by glendaviesnz** [comment]
Will start taking a look at this, but happy for others to also jump in and help.

**Comment by jameskoster** [comment]
@glendaviesnz I think @jorgefilipecosta has made a start on this so perhaps worth syncing up.

**Comment by glendaviesnz** [comment]
@jameskoster is there a reason why we are not offering the choice of template/pattern as part of the first dialog where the name is entered?

<img width="572" alt="Screenshot 2022-11-25 at 12 50 31 PM" src="https://user-images.githubusercontent.com/3629020/203875600-17f3f708-9519-40da-8980-dc39a5930e58.png">

If this dialog is expanded to include the choice of templates, or blank, it saves presenting a second dialog after the template is created

**Comment by jameskoster** [comment]
We may do eventually, but I think the aim is to get these two very similar flows (choosing a pattern while creating a page & choosing a pattern while creating a template) aligned first. Since the former already exists it makes sense to reuse what we have.

**Comment by jorgefilipecosta** [comment]
Given that we are implementing issue https://github.com/WordPress/gutenberg/issues/42325 where the user chooses a pattern in a modal before even opening the sie editor for the template. I think we should close this issue as https://github.com/WordPress/gutenberg/issues/42325 already accomplishes the objective of suggesting a pattern to the user and a template never starts blank unless the user explicitly asked for the blank one.

**Comment by glendaviesnz** [comment]
@jameskoster I agree with @jorgefilipecosta that this issue doesn't make sense now with the options already presented in the 'Add New' listing.

@jorgefilipecosta, in terms of the `blank` option, is the plan for the `Custom template` option to return blank, or are you adding an explicit `Blank` option to the menu?

**Comment by jameskoster** [comment]
The problem is more about divergent executions of the same UX. We don't really want to put ourselves in a position where – when we want to update the pattern selection UX – we need to consider two different implementations.

The main issue with the designs originally shared in #42325 is that they were entirely template centric. We need to think a bit more about the page/post creation flow as well, and ensure we come up with a solution that can serve both flows.

> this issue doesn't make sense

To clarify: this issue is purely about displaying patterns (+ fallback / start blank) when a template is created. #42325 should concentrate on making the "add new template" menu scale better to accommodate all the different template types we added recently.

**Comment by glendaviesnz** [comment]
> this issue is purely about displaying patterns (+ fallback / start blank) when a template is created. 

I am not sure that I understand why you would show these options when the user has either chosen a template type from the  `add new template` menu, or chosen the custom option which I assume would be blank - seems like we are forcing them to decide twice? ... but I may be overlooking something - maybe a brief video showing the relation of the flow outlined here and how it relates to the `create new template` menu might help?

**Comment by jameskoster** [comment]
Because choosing the template type is only the first step, after which you still need to design the template. A 404 template for example can have endless configurations, which is why we present patterns.

The patterns are contextual to the template you chose. 

> custom option which I assume would be blank

No, the custom template can be applied to any single post or page, so the patterns we suggest will all be different ways to display a single post / page.

**Comment by jorgefilipecosta** [comment]
> Because choosing the template type is only the first step, after which you still need to design the template. A 404 template for example can have endless configurations, which is why we present patterns.

So on the add new template the user chooses the template type e.g: 404, or single post and chooses a pattern for that template (or blank if the user really wants blank) the site editor loads the new created template with the choose pattern. In which case we would show another modal to choose a pattern? The user already choose a pattern before.
If the idea is really to have two modals and choose patterns two times, for the 404 template type could you given examples of patterns the user would choose on the add new template modal, and the patterns the user would choose after the site editor loads? Are they the same patterns?

**Comment by jameskoster** [comment]
No, the user should not be choosing the pattern twice 😭

Template selection and pattern selection are separate parts of the same overall flow. That is:

1. Click Add New template
2. Choose the template e.g. Page
3. Choose whether to create the generic `page` template, or to create a template for a specific page
4. Select the specific page you want to create the template for
5. You're sent to the Editor and a modal appears containing patterns for displaying a page

This issue concentrates _purely on that fifth step_. Steps 1 through 4 need more design work. Sorry for the confusion here.

**Comment by ndiego** [comment]
@jameskoster @jorgefilipecosta Just checking in to see if we still aiming for 6.2 with this? Thanks!

**Comment by jameskoster** [comment]
The first task was merged last week (#46248), and I think that can be included in 6.2. 

@jorgefilipecosta can confirm, but I don't believe the second task was factored into the work in that PR, so we probably need to leave the issue open.

**Comment by ndiego** [comment]
@jameskoster https://github.com/WordPress/gutenberg/pull/46248 was included in Gutenberg 15.2 (not included in 6.2), and I do not believe it was backported to the 6.2 Betas. @Mamaduka @ntsekouras are you able to confirm this?

**Comment by Mamaduka** [comment]
Gutenberg 15.2 was released yesterday; we need more time to test #46248 to be included in the release IMO. So I would recommend punting this for WP 6.3.

**Comment by ntsekouras** [comment]
This is punted already since: https://github.com/WordPress/gutenberg/pull/48077, that moved the files in `compat/6.3`.

**Comment by jameskoster** [comment]
Dang, that's a shame. 46248 softened a number of rough edges in the template creation flow 😓

**Comment by SaxonF** [comment]
> You're sent to the Editor and a modal appears containing patterns for displaying a page

@jameskoster do you think this should be part of the new drill-down experience vs a modal? Another option is just opening inserter in editor as this helps from an educational aspect as well. This is where some guidelines around site editor sidebar vs editor would help us decide where an experience like this should live. 

![image](https://user-images.githubusercontent.com/1072756/228099760-4ecb3156-63ef-45e7-9fca-27bbb5185753.png)

**Comment by jameskoster** [comment]
> @jameskoster do you think this should be part of the new drill-down experience vs a modal? 

Yes, we should explore that. 

> Another option is just opening inserter in editor as this helps from an educational aspect as well.

Maybe it's reachable in both places 🙊 The 'full-page' patterns will be available in the Inserter regardless. And surfacing them in the dark material could be nice as well, because it's not beyond the realms of possibility that it'd enable you to create a template without having to invoke the full screen editor at all.

**Comment by jameskoster** [comment]
@SaxonF are you seeing the "Apply template to" dropdown containing the options from this menu?

<img width="250" alt="Screenshot 2023-03-29 at 19 17 59" src="https://user-images.githubusercontent.com/846565/228631380-a5691137-4bed-4a73-8947-646a23620843.png">

I'm curious how we'd present 'custom' templates.

The starting points will update as you work through the form so we might need a loading state. Could be worthwhile to quickly prototype this one.

**Comment by SaxonF** [comment]
> are you seeing the "Apply template to" dropdown containing the options from this menu?

Previously shared this somewhere else so obviously doesn't reflect dark mode but the idea is that rather than trying to explain the complex template hierarchy in a single dropdown we take a more natural language approach by splitting into two inputs. custom templates would be the bottom option of the first input.

https://user-images.githubusercontent.com/1072756/228707717-0f264d9f-7494-495b-ac89-f275549832d1.mp4

**Comment by jameskoster** [comment]
Yeah, it's quite tricky to support the full expressiveness of the template hierarchy and present it in a way that feels intuitive :) 

I like the natural language, but I'm missing ways to create a template for a post tag or category, or for a specific author. 

Another idea to avoid overwhelming in that first dropdown could be to center its contents around post types, and reveal subsequent dropdowns based on the initial selection, kind of like a filter:

<img width="855" alt="Filter" src="https://user-images.githubusercontent.com/846565/228792799-32be87d2-e544-4b46-9fdf-4f28cc289ce0.png">

It seems that whichever way we slice it, it comes down to: Fewer dropdowns with longer menus. More dropdowns with shorter menus.

**Comment by SaxonF** [comment]
I like that. To me it feels more intuitive as it focuses more on the outcome vs trying to gauge the outcome from template name + description. We can also be clever about certain selections e.g. decide whether to use home or front-page based on site configuration.

**Comment by jameskoster** [comment]
One other thing that came up in a discussion with @jasmussen today was a need for a more comfortable way to browse many starting points. Perhaps something like the pattern directory modal?

**Comment by jasmussen** [comment]
This is the mockup we discussed:

<img width="1596" alt="Editor" src="https://user-images.githubusercontent.com/1204802/229055162-d39ea1cf-f5be-4039-bf21-afb82d0b0080.png">

Note that I don't consider it a particularly good design: the "Start blank" could arguably be a blank page inside, the modal is a bit big and unstyled, etc.  — but it's spacious. 

Even if we do keep things in the drilldown detail pages, which I think brings with it a lot of solid benefits, a modal like this could potentially be an extra affordance. Maybe.

**Comment by jameskoster** [comment]
Yup that's what I meant: an additional affordance, similar to the "Explore all patterns" button in the Inserter 👍

**Comment by jameskoster** [comment]
There have been conversations about moving the pattern selection process to the sidebar, which is a bit related to https://github.com/WordPress/gutenberg/issues/50739. 

But since pattern suggestions are now displayed (both fall backs and contextual), I think we can close this issue.

<img width="1308" alt="Screenshot 2023-05-18 at 18 57 57" src="https://github.com/WordPress/gutenberg/assets/846565/cdb10f9c-3f42-486f-bf03-0ad45cd52fc6">
